devito/types/separable.py

```diff
@@ -0,0 +1,221 @@
+
+    def __getitem__(self, dim):
+        return self.subfunctions[dim]
+
+    @property
+    def data(self):
+        return tuple(f.data for f in self.subfunctions.value)
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
+        grid = kwargs.get('grid')
+        dimensions = kwargs.get('dimensions')
+        if grid is None:
+            if dimensions is None:
+                raise TypeError("Need either `grid` or `dimensions`")
+        elif dimensions is None:
+            dimensions = grid.dimensions
+        return dimensions
+
+    @property
+    def args(self):
+        return self._args
+
+
+class SeparableTimeFunction(SeparableFunction):
+    """
+    A TimeFunction separable in its Dimensions, ie f(t,x,y,z) = f(t)*f(x)*f(y)*f(z)
+
+    Take the same parameters as a TimeFunction plus
+
+    Parameters
+    ----------
+    sep: str
+        Type of separation as a product or sum of Function, 'sum' or 'prod'.
+    separated: Dimension or tuple of Dimensions
+        Dimension in which the TimeFunction is separable.
+
+    Examples
+    --------
+    Creation
+
+    >>> from devito import Grid, TimeFunction
+    >>> grid = Grid(shape=(4, 4, 5))
+    >>> x, y, z = grid.dimensions
+    >>> t = grid.stepping_dim
+    >>> SeparableTimeFunction(name='a', grid=grid, space_order=8, time_order=2)
+    a_t(t)*a_x(x)*a_y(y)*a_z(z)
+    >>> SeparableTimeFunction(name='b', grid=grid, sep='sum')
+    b_t(t) + b_x(x) + b_y(y) + b_z(z)
+    >>> SeparableTimeFunction(name='c', grid=grid, sep='sum', separated=t)
+    c_t(t) + c_xyz(x, y, z)
+    >>> SeparableTimeFunction(name='d', grid=grid, separated=(t, x))
+    d_t(t)*d_x(x)*d_yz(y, z)
+    """
+    _time_position = 0
+
+    def __init__(self, *args, **kwargs):
+        if not self._cached():
+            self._time_order = kwargs.get('time_order', 1)
+            super(SeparableTimeFunction, self).__init__(*args, **kwargs)
+
+    @property
+    def time_order(self):
+        return self._time_order
+
+    def __setup_dimensions__(self, **kwargs):
+        # Is it sum or product separable
+        save = kwargs.get('save')
+        time_dim = self.grid.time_dim if isinstance(save, int) else self.grid.stepping_dim
+        sep = as_tuple(kwargs.get('separated', (time_dim,) + self.grid.dimensions))
+        nonsep = tuple(d for d in self.grid.dimensions if d not in sep)
+        return sep, nonsep
+
+    def function(self, dimensions, **kwargs):
+        dims = as_tuple(dimensions)
+        func_type = TimeFunction if any(d.is_Time for d in dims) else Function
+        names = "".join(d.name for d in dims)
+        save = kwargs.get('save', None)
+        shape = tuple(save or self.time_order+1 if d.is_Time
+                      else self.grid.dimension_map[d].loc for d in dims)
+        kwargs["name"] = '%s_%s' % (self.name, names)
+        func = func_type(dimensions=dims, shape=shape, **kwargs)
+        return func
+
+    @classmethod
+    def __indices_setup__(cls, **kwargs):
```

Comment:
this is a copy-paste, and therefore it's not acceptable

Reply:
This has to go here for debug, same for other one